## Question 1
What's the version of Pandas that you installed?

You can get the version information using the __version__ field:

In [7]:
import pandas as pd
pd.__version__

'2.2.2'

### Getting the data
For this homework, we'll use the California Housing Prices dataset. Download it from here.

In [8]:
!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv

--2024-08-06 16:55:39--  https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1423529 (1.4M) [text/plain]
Saving to: ‘housing.csv’

housing.csv         100%[===================>]   1.36M  3.91MB/s    in 0.3s    

2024-08-06 16:55:40 (3.91 MB/s) - ‘housing.csv’ saved [1423529/1423529]



In [41]:
df = pd.read_csv('../housing.csv')
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

## Question 2
How many columns are in the dataset?

In [10]:
len(df.columns)

10

## Question 3
Which columns in the dataset have missing values?

In [42]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

`total_bedrooms` have 207 missing values

## Question 4
How many unique values does the ocean_proximity column have?

In [14]:
df['ocean_proximity'].nunique()

5

## Question 5
What's the average value of the median_house_value for the houses located near the bay?

In [17]:
df['ocean_proximity'].unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [25]:
df_near_bay = df[df['ocean_proximity'] == 'NEAR BAY']
df_near_bay['median_house_value'].mean()

np.float64(259212.31179039303)

## Question 6
1. Calculate the average of total_bedrooms column in the dataset.
2. Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
3. Now, calculate the average of total_bedrooms again.
Has it changed?

In [43]:
mean_bedrooms=df['total_bedrooms'].mean()
print(mean_bedrooms)

537.8705525375618


In [44]:
df['total_bedrooms'] = df['total_bedrooms'].fillna(mean_bedrooms)

In [45]:
df.isnull().sum()

longitude             0
latitude              0
housing_median_age    0
total_rooms           0
total_bedrooms        0
population            0
households            0
median_income         0
median_house_value    0
ocean_proximity       0
dtype: int64

In [51]:
df.shape[0]
df['total_bedrooms'].mean()

np.float64(537.8705525375617)

## Question 7
1. Select all the options located on islands.
2. Select only columns housing_median_age, total_rooms, total_bedrooms.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [950, 1300, 800, 1000, 1300].
7. Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
8. What's the value of the last element of w?

In [57]:
df_islands = df[df['ocean_proximity'] == 'ISLAND']
df_islands.shape[0]

5

In [55]:
df_selected_columns = df_islands[['housing_median_age', 'total_rooms', 'total_bedrooms']]
df_selected_columns

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [59]:
X = df_selected_columns.values

In [60]:
X_transpose = X.T
X_transpose

array([[  27.,   52.,   52.,   52.,   29.],
       [1675., 2359., 2127.,  996.,  716.],
       [ 521.,  591.,  512.,  264.,  214.]])

In [62]:
XTX = X_transpose.dot(X)

In [67]:
import numpy as np

XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [69]:
y = np.array([950, 1300, 800, 1000, 1300])

In [70]:
XTX_inv_XT = XTX_inv.dot(X_transpose)
w = XTX_inv_XT.dot(y)
w

array([23.12330961, -1.48124183,  5.69922946])

In [75]:
print(f"What's the value of the last element of w?, {w[2]}")

What's the value of the last element of w?, 5.6992294550655735
